### ISSUE TITLE
spy shows the matrix transposed

### ISSUE THREAD

**Issue by dpsanders** [tbreloff/Plots.jl#196]
An upper-triangular matrix is shown as lower-triangular by `spy`.

**Comment by tbreloff** [comment]
I agree... there needs to be a transpose somewhere, but I need to figure out the correct place to add it.

**Comment by dpsanders** [comment]
Yes, I see -- there seems to be a similar flipping issue with `heatmap`, the usual problem of how to treat "axes" in a matrix, which is certainly tricky since in different situations you might want different solutions. (I guess this is what the `yflip` keyword argument is about.)

In the case of `spy`, though, it seems pretty unambiguous that you would want the image on screen to correspond to the layout of the matrix as shown by Julia, so for now you could just add the transpose in the definition of `spy`:

```julia
function spy{T<:Real}(z::AMat{T}; kw...)
  heatmap(map(zi->float(zi!=0), z)'; leg=false, yflip=true, kw...)
end
```

**Comment by tbreloff** [comment]
Yes I did that just now, but just as you said it might be better to do the transpose for all heatmaps.  And maybe also contours, surfaces, etc?  I'm wondering if my `Plots.Surface` should really be a subtype of AbstractMatrix and make it effectively a "transpose view".  Certainly adding a transpose to spy will fix this particular problem, but I want to make sure I have the correct behavior generally.

Note: you can always pass in a transposed matrix to the current function to get it working now...

**Comment by tbreloff** [comment]
And if it's not clear... I'd love to have some other eyes looking at the current behavior of heatmap, contour, etc to figure out what needs to be transposed/flipped (and under what circumstances).  Help appreciated!

**Comment by dpsanders** [comment]
Hmm I see. I guess this is a problem that all plotting libraries have previously battled with. Is there any kind of consensus out there? I certainly remember being frustrated with PyPlot's take in the past, but don't really remember details. Opposite conventions for `imshow` and `pcolor`?

**Comment by tbreloff** [comment]
For reference: 

http://stackoverflow.com/questions/21166679/when-to-use-imshow-over-pcolormesh
http://thomas-cokelaer.info/blog/2014/05/matplotlib-difference-between-pcolor-pcolormesh-and-imshow/

**Comment by tbreloff** [comment]
From the [matplotlib docs](http://matplotlib.org/api/pyplot_api.html#matplotlib.pyplot.pcolor):

> The grid orientation follows the MATLAB convention: an array C with shape (nrows, ncolumns) is plotted with the column number as X and the row number as Y, increasing up; hence it is plotted the way the array would be printed, except that the Y axis is reversed. That is, C is taken as C*(*y, x).

`imshow` is similar except that the y-axis is flipped.

These conventions are contrary to the way that I, personally, would think about plotting a surface/contour.  The first dimension of a matrix is rows, which should (IMO) correspond with the first dimension of the plot (the x-axis).

I want to be able to do this:

```julia
x = linspace(0,3,20)
y = linspace(0,10,20)
z = Float64[5xi+yi for xi=x,yi=y]
heatmap(x, y, z)
```
so that the x/y line up with how they were constructed.

I have to pick a convention, and there's clearly not a 100% right answer.  I'm happy to listen to opinions/arguments on the matter, but I'm not sure I want to change the behavior.

**Comment by tbreloff** [comment]
Before I forget... I had an idea to add a keyword 'match_dimensions' with alias 'transpose_z', for 2 cool benefits:

- The user can choose her preferred convention... Plots-style or Matplotlib-style
- we can avoid a double-transpose for 'spy'...  Just negate the 'match_dimensions' argument!

**Comment by jebej** [comment]
Since the goal is the display of a matrix, it makes more sense to have the plot have the same orientation as the matrix itself, despite this meaning that rows will map to the y axis. The y axis should also be inverted such that the 0,0 entry is a the top left.

**Comment by tbreloff** [comment]
Yes I'm in agreement about what spy should look like (it should look like a
matrix), but I dont think heatmap should have the same logic. I also don't
think there's inherently a right answer here, which is why I suggested
making it a keyword, which defaults to the current behavior.   You can
always change the default if you want to see the transposed version like
matplotlib.

**Comment by jebej** [comment]
What do you mean when you say you don't think heatmaps shoule have the same logic?

I was trying to say that heatmaps are not matrices (at least they should not be, by default) and so the x=0,y=0 coordinate should be in the bottom left of the plot.

**Comment by tbreloff** [comment]
Yup we agree there. Spy (0,0) should be top-left corner, and heatmap (0,0)
should be bottom-left. The open question is whether the first dimension of
the z matrix (rows) should correspend to the x axis or the y axis.
Matplotlib chose y (like an upside-down matrix)... I think x feels much
more intuitive/natural. Maybe it's just me... If there's agreement from the
community about the best convention, I'm happy to default to the crowd.

**Comment by tbreloff** [comment]
Yup we agree there. Spy (0,0) should be top-left corner, and heatmap (0,0)
should be bottom-left. The open question is whether the first dimension of
the z matrix (rows) should correspend to the x axis or the y axis.
Matplotlib chose y (like an upside-down matrix)... I think x feels much
more intuitive/natural. Maybe it's just me... If there's agreement from the
community about the best convention, I'm happy to default to the crowd.

**Comment by jebej** [comment]
Like I said above, I think it makes more sense to keep the orientation the same as if you were writing down the matrix. Rows of the matrix should thus be rows on the plot.

Matplotlib does not plot the matrix upside down (see [here](http://matplotlib.org/examples/pylab_examples/spy_demos.html), it plots it, well, like the matrix. It seems more natural to me this way.

Maybe it will will help to remember a 1D array is a column vector, and so the index corresponds to the row. Same with a 2D array, the first index corresponds to the row, and the second to the column.

Having the first index correspond to x would mean the matrix would be 90 degrees rotated.

**Comment by tbreloff** [comment]
So I just ran through the matplotlib and plotly examples... matplotlib's pcolor and plotly's heatmap both have (0,0) in the lower-right and the rows (first dimension) correspond to the y-axis for both.  This looks like the z matrix is just flipped upside-down.

Since 2 major packages have accepted that convention, I guess I can follow the crowd (as much as I hate to do it).  I think the behavior of `spy` is currently correct (i.e. all non-zero values are the same color), but I will probably change the default colormap to match the background color for zeros, and match the foreground color for non-zeros.

As for Gadfly, I currently don't support heatmaps in Gadfly because I didn't realize the function existed, but they just named heatmaps spy!  So I'll add support for both heatmap and spy in Gadfly.

The todo list:

- [ ] add keyword `match_dimensions` (alias ideas welcome!) for mapping rows to x (true) or y (false) -- default value is false, so rows will be the y-axis, like an upside-down matrix (same as matplotlib)
- [ ] force `match_dimensions=false` for spy (rows are y-axis)
- [ ] add default color logic for spy (will usually be `ColorGradient([bgcolor, fgcolor])`)
- [ ] implement heatmaps in gadfly by calling `Gadfly.spy` (if I can get it to work)
- [ ] ensure backends use match_dimensions in heatmap logic (pyplot, plotly, plotlyjs, gr)

Anything else I'm forgetting?  This will be breaking... need to figure out what might break...

**Comment by tbreloff** [comment]
Note: need to ensure heatmap, surface, wireframe, and contours are consistent, as they all represent the same concept.

**Comment by jebej** [comment]
> So I just ran through the matplotlib and plotly examples... matplotlib's pcolor and plotly's heatmap both have (0,0) in the lower-right and the rows (first dimension) correspond to the y-axis for both. This looks like the z matrix is just flipped upside-down.

Yes, but I think you meant the 0,0 is in the lower-left.

The way I see it, we should have three functions:

TL;DR: `spy`, `imagesc` and `heatmap`, all three have the rows on the y axis, and the columns on the x axis. Only `heatmap` is the upside-down matrix, with 0,0 in the bottom left.

Let's take as an example the following matrix:
```
[0 4 7;
 2 5 8;
 3 0 9]
```
1. `spy` which plots the sparsity matrix as a binary, this would mean that on screen we would have:
![spy](https://cloud.githubusercontent.com/assets/11141558/15004200/13d316f0-1182-11e6-9c57-888c3bb70182.png)
Notice the orientation is the same as the actual matrix.

2. some function to plot the actual matrix (matlab has `imagesc`), the only difference with spy is that now the values are plotted with a colormap.
![imagesc](https://cloud.githubusercontent.com/assets/11141558/15004226/61111746-1182-11e6-8ef8-453f18d57b7b.png)

3. `heatmap` which would plot the values colored like the above, but now the matrix is upside down. This corresponds to the matplotlib `pcolor` and plotly `heatmap`, as well as Gadfly [rectbin](http://gadflyjl.org/geom_rectbin.html)
![](http://matplotlib.org/_images/pcolor_small.png)

**Comment by jebej** [comment]
I did all three figures in Matlab, with the `imagesc` function, reversing the y-axis manually for the heatmap (figure 3). The default output is given by figure 2.

**Comment by tbreloff** [comment]
Oops yes bottom-left. I think we're on the same page. Does anyone else want
to chime in?

**Comment by diegozea** [comment]
I agree with @jebej about `spy` being different from `image...` and they different in orientation from `heatmap`.

**Comment by pkofod** [comment]
Just to be sure, are we considering adapting `imagesc` as a name? It's one of those typical matlab abbreviations that makes no sense. I never knew it was "image scaled colors`, but I feel like the functionality is something that should be handled by a color map keyword, not the line type. 

I agree with the orientation presented above.

**Comment by diegozea** [comment]
I didn't know that *sc* was for *scaled colors*.

**Comment by pkofod** [comment]
Apparently, the idea is that the lowest color in the color map gets mapped to the lowest value in the matrix, and vice versa for the highest value of the color map (so the color map is scaled to match the min and max of the matrix elements). I just don't get why that isn't simply an option rather than part of the name.

**Comment by lobingera** [comment]
Just for completeness:
>> help imagesc
 imagesc Scale data and display as image.
    imagesc(...) is the same as IMAGE(...) except the data is scaled
    to use the full colormap.
    
    imagesc(...,CLIM) where CLIM = [CLOW CHIGH] can specify the
    scaling.
 
    See also imshow,image, colorbar, imread, imwrite.

>> help spy
 spy Visualize sparsity pattern.
    spy(S) plots the sparsity pattern of the matrix S.
 
    spy(S,'LineSpec') uses the color and marker from the line
    specification string 'LineSpec' (See PLOT for possibilities).
 
    spy(S,markersize) uses the specified marker size instead of
    a size which depends upon the figure size and the matrix order.
 
    spy(S,'LineSpec',markersize) sets both.
 
    spy(S,markersize,'LineSpec') also works.

I think there are two options: re-use matlab names, as they are common OR invent clear,new,descriptive names.

**Comment by tbreloff** [comment]
I hate the imagesc name. Don't worry... No plans to add it. I have thought
that we need an imshow/imagesc type of function though. I might be
convinced to add imshow as a convenience wrapper around 'heatmap(args...;
kw..., yflip=true)', and also that plotting a Images.image should call this
method.

**Comment by Evizero** [comment]
>  I might be convinced to add imshow as a convenience wrapper around 'heatmap(args...; kw..., yflip=true)', and also that plotting a Images.image should call this method.

:+1:  for conveniently displaying images using Plots

**Comment by tbreloff** [comment]
Thinking about this more, maybe we do need a dedicated series type for
images (or image-like data). I can't think if a good way to map a single z
value to an RGB tuple.

Name bikeshedding please... Should we go wih imshow or something else?

**Comment by KristofferC** [comment]
`imageshow` then IMHO.

**Comment by pkofod** [comment]
I  may very well be missing something obvious but why not just image? It's not scattershow or lineshow either.

**Comment by lobingera** [comment]
short function names like 'image' most likely interfere with other exported names.

**Comment by tbreloff** [comment]
I thought it was uses for image creation by Images. I'll need to check
that.  If it's available, I do like the consistency of 'image'... The
concern is namespace clashes.

The alternative is that there's no method for images, but the series type
is image. So for Image img, matrix z:

plot(img)
plot(z, t=:image)

Maybe imageplot(z)?

**Comment by Evizero** [comment]
if a custom function name can be avoided without ugly sideffects, then I think just `plot(img)`, where `typeof(img) <: Image`, would be the prettiest and most intuitive solution, given that Plots.jl follows the one-function-to-rule-them-all approach

**Comment by Evizero** [comment]
To be a bit more transparent. The main thing I would hope to be able to do is abuse `subplot` to conveniently plot multiple images in one go with the ability to give each one a title.

**Comment by tbreloff** [comment]
> I would hope to be able to do is abuse subplot to conveniently plot multiple images in one go

Use it and abuse it... it's all good.

Any convenience method will just be a wrapper around plot:

```julia
imageplot(args...; kw...) = plot(args...; kw..., seriestype = :image)
```

**Comment by jebej** [comment]
> we need an imshow/imagesc type of function

imshow is pretty different from imagesc

> Thinking about this more, maybe we do need a dedicated series type for
images (or image-like data). I can't think if a good way to map a single z
value to an RGB tuple.

There isn't one without making assumptions. this is what Matlab does for imshow:

```
imshow(I)
I — Input image
scalar | vector | matrix | m-by-n-by-3 array
Input image, specified as a scalar, vector, or matrix representing a grayscale, RGB, or binary image. Multi-plane image inputs must be RGB images of size m-by-n-by-3. An RGB image can be uint8, uint16, single, or double. A grayscale image can be any numeric data type. A binary image is of class logical.
```

Thus when a single matrix is given, Matlab assumes they represent a grayscale (I think the elements must be between 0 and 1).

**Comment by tbreloff** [comment]
Sorry I actually said it backwards... I mean that I can't fit an RBG image into the heatmap/colormap concept.  I need to be able to pass the 3-dimensional array directly.

**Comment by jebej** [comment]
> add imshow as a convenience wrapper around 'heatmap(args...;
kw..., yflip=true)'

I think that the convenience wrapper would be `imagesc` in this case, since a reverse heatmap would look like figure 2 above.

**Comment by jebej** [comment]
> Sorry I actually said it backwards... I mean that I can't fit an RBG image into the heatmap/colormap concept. I need to be able to pass the 3-dimensional array directly.

In my opinion, plotting an image should not need a colormap.

**Comment by tbreloff** [comment]
> plotting an image should not need a colormap

Right... that's what I meant.

I'm partially thinking about API here, and partially internal implementation.  If I can shoehorn this into the heatmap type, then anything that implements a heatmap will automatically implement spy, imageplot, etc.  If it's a totally distinct plotting pipeline then it's more effort to get all the backends working.

**Comment by jebej** [comment]
I think the API would be okay with a `plot(args...; kw..., seriestype = :image)` function, as well as a wrapper (`imageshow` or `imshow`, or even `image` if possible). This would match Matlab and matplotlib, which I think is important.

Regarding the backend, I don't think you will be able to use heatmap to display an image, you will have to use the appropriate image display functions like `imshow` in matplotlib. I'm sure GR has the option to do that as well.

It might be possible to do the same in Gadfly, but you would probably need to dig in deep. It could not be worth the effort for now, as you would not be using the public gadfly API and it could get changed. I cannot see how to do that in Plotly.

**Comment by jebej** [comment]
> If I can shoehorn this into the heatmap type

You could definitely use the heatmap type by mapping every RGB triplet to a single number (ie the product space RxGxB) and generate a colorscale with the matching colors (0,0,0),(0,0,1)... until (255,255,255), but that seems a little bit hackish. It would also not be easily extensible to display images in other colorspaces (which we might not care about).

If we only want a simple way to display an RGB image that might be enough. I figure that if someone wanted to do something more sophisticated they would use an other package.

**Comment by tbreloff** [comment]
> but that seems a little bit hackish

Yes this is the sort of hack I'm thinking about.  Not ideal, but it could be a simple design for those backends that either don't support images, or we just haven't implemented them yet:

```julia
# in the plotting code:
if d[:linetype] == :image && !(:image in supportedTypes())
    d[:linetype] = :heatmap
    d[:z], d[:fillcolor] = replace_image_with_heatmap(d[:z])
end
```

Then for backends that support images natively it can be handled natively.

**Comment by tbreloff** [comment]
The hack is working (sort of... much better for grayscale), which maps matrices of colors (images) into a heatmap with a custom colormap for that image.  It sucks, as compared to native image support within a backend, but it's something.

For PyPlot, I implemented a pass-through to imshow, so you can just `plot(img)` an image, and it flows through the normal Plots pipeline (it can be part of a more complex plot).  Here's an example:

![tmp](https://cloud.githubusercontent.com/assets/933338/15125976/c3339276-15fd-11e6-840b-a58425cced23.png)

**Comment by jebej** [comment]
What method names/plot types are we converging towards?

It seems you are now assuming that if someone write `plot(x)` where `x` is an m-by-n-by-3 array they want to display an image. That seems reasonable to me (ie I can't think of a situation where that wouldn't be the case).

Regarding more regular matrix plotting, where `x` is an m-by-n array, would it be useful to have both a `heatmap` function, with (0,0) in the bottom left, and a convenience wrapper function that would plot the matrix with (0,0) in the top left (like you described above)?

**Comment by tbreloff** [comment]
I think non-standard plotting routines should be implemented by recipes of some form... Either "type recipes" like for DataFrames or Distributions, or "series recipes" like violin and quiver plots. (I'm still nailing down terminology). So a flipped heatmap is nothing more that... A flipped heatmap. I don't think it needs special handling, but you could always define a shorthand if you want. Just keep in mind this is different than imshow, which displays a matrix of color values.

**Comment by tbreloff** [comment]
I'm closing this in favor of #273.  Plotting images is supported now, though the "heatmap hack" should be changed to a series recipe...
